### ISSUE TITLE
Unable to use plugin with Postgres schema other than 'public'

### ISSUE THREAD

**Issue by alamote** [dereuromark/cakephp-queue#367]
Assuming there is a connection in app.php like 

```
'app' => [
    'className' => Connection::class,
    'driver' => Postgres::class,
    'schema' => 'app',
     ... 
],
```

and we are using this connection in app_queue.php

```
'connection' => 'app',
```

It's expected that plugin will use specified schema to store queued_jobs and queue_processes tables.
But when task is processed, plugin tries to update table without schema prefix ('public' schema is used in this case) and expectedly, an error appears that there is no such table (it's located in 'app' schema).

Solution:

Redeclare `setTable` method in `QueuedJobsTable` and `QueueProcessesTable` models like this:

```
public function setTable(string $table)
{
        $parts = [$table];
        if ($this->getDriverName() === static::DRIVER_POSTGRES) {
            array_unshift($parts, $this->getConnection()->config()['schema'] ?? 'public');
        }
        $this->_table = implode('.', $parts);

        return $this;
}
```

**Comment by dereuromark** [comment]
Can you make a PR to show your changes?
I dont quite fully get it.

**Comment by dereuromark** [comment]
It does feel kind of hacky though
I wonder if there isnt a more "framework appropriate" approach to it..

cc @ADmad

**Comment by alamote** [comment]
@dereuromark But CakePHP's doc for setTable method says:

```
This can include the database schema name in the form 'schema.table'. If the name must be quoted, enable automatic identifier quoting.
```

So not sure if it's not "framework appropriate" approach

**Comment by LordSimal** [comment]
I am unable to reproduce this problem in CakePHP's tests as well as in this plugin's tests using a custom `app` schema as a DB. Are you certain this is not something you have on your side?

This seems very strang because according to your description

> But when task is processed, plugin tries to update table without schema prefix

this would mean, that this ONLY happens when the worker is running and updating the job.

Also: Why woult it be able to fetch a job from `queue_jobs` but not update it? This is not consistent.

**Comment by alamote** [comment]
@LordSimal My understanding is that this happens when there are multiple connections in the app. Firstly, Queue Processor is working with connection that is specified in app_queue.php, then some job is using another connection and after that QueuedJobsTable's markJobDone method fails.

I'm getting such excepltion in QueuedJobsTable:687:

```
SQLSTATE[42P01]: Undefined table: 7 ERROR:  relation "queued_jobs" does not exist
LINE 1: UPDATE queued_jobs SET completed = $1 , progress = $2 WHERE ...
               ^
```

**Comment by dereuromark** [comment]
Can you try not to use app_queue config here, but instead use your default config?
Note how in next major it will go away with this file (https://github.com/dereuromark/cakephp-queue/blob/cake5/config/app.example.php) in favor of a example file only.

**Comment by alamote** [comment]
@dereuromark not sure how this can help. If I'm not using app_queue.php (don't configure Queue) then `Cannot describe queue_processes. It has 0 columns.` error appears (what is expected because there are no queue tables in the default schema). Moving Queue config to app.php (`Queue` key in return array) leads to the same result when it was in app_queue.php

**Comment by LordSimal** [comment]
Since we are working with at least 2 different postgres schemas and/or DBs can you try and get a theoretical example working here?

Somehing like:

## Connection 1: `public`
Postgres schema: `public`
Contains tables `posts`, `post_meta` etc.

## Connection 2: `default`
Postgres schema: `app`
Contains the `queued_jobs` and `queue_processes` tables

The worker run with the configuration associated to Connection 2.
But a job being processed by that worker tries to fetch data from Connection 1 via....

**Comment by LordSimal** [comment]
@alamote what @dereuromark means, is that in the past there was a `connection` config inside the `app_queue.php` file as you can see here.
https://github.com/dereuromark/cakephp-queue/blob/cake3/config/app_queue.php#L44-L50
But this is not recommended anymore so **if you use it, try removing it** and see if this solves your problem

**Comment by LordSimal** [comment]
I missunderstood dereuromark there, the `connection` key is fine there so this shouldn't be the problem.
Will have to think about that for a bit what may be the problem here.

**Comment by LordSimal** [comment]
Do you by chance have any Connection configuration present inside your `config` directory which specifies no `schema`?

The error `UPDATE queued_jobs SET completed = $1 , progress = $2 WHERE` indicates that a table instance used a connection config which doesn't declare a `schema`.

You could also try and add some debug information just bellow that `->save()` call which fails to get some more information on what config it used.

**Comment by alamote** [comment]
@LordSimal all connections have a schema name:
<img width="1026" alt="Screenshot 2023-02-14 at 18 46 14" src="https://user-images.githubusercontent.com/9486443/218802251-5174fd5d-b8cf-47c9-926d-ec321d720c7b.png">
<img width="1026" alt="Screenshot 2023-02-14 at 18 46 39" src="https://user-images.githubusercontent.com/9486443/218802357-e44dc523-8e67-4a6d-aeee-180f781f8dd9.png">
<img width="1026" alt="Screenshot 2023-02-14 at 18 46 54" src="https://user-images.githubusercontent.com/9486443/218802422-c12f3d29-86a7-4239-9341-42a18a486ce4.png">

Here is a `$this->_connection->config()` output before `$this->save($job)`
<img width="1026" alt="Screenshot 2023-02-14 at 18 43 09" src="https://user-images.githubusercontent.com/9486443/218801496-7c253599-4f96-4d9f-91aa-a5078b282a7f.png">

**Comment by alamote** [comment]
@LordSimal all connections have a schema name:

```
result = {array[15]} 
 className = "Cake\Database\Connection"
 driver = "Cake\Database\Driver\Postgres"
 name = "default"
 schema = "cr"
...
```

```
result = {array[15]} 
 className = "Cake\Database\Connection"
 driver = "Cake\Database\Driver\Postgres"
 name = "reminders"
 schema = "reminders"
...
```

```
result = {array[15]} 
 className = "Cake\Database\Connection"
 driver = "Cake\Database\Driver\Postgres"
 name = "app"
 schema = "app"
...
```

Here is a `$this->_connection->config()` output before `$this->save($job)`

```
result = {array[14]} 
 driver = "Cake\Database\Driver\Postgres"
 name = "app"
 schema = "app"
...
```

**Comment by alamote** [comment]
@dereuromark @LordSimal I think I can reproduce the issue in clear cake app (using docker for simplicity) later, will push and notify when it's ready

**Comment by alamote** [comment]
@dereuromark here is an update: strange thing happens - I created a similar app and trying to reproduce the issue, but it works as expected. But when I copied the same task to my project, it fails... I continue to investigate...

**Comment by alamote** [comment]
@dereuromark @LordSimal found the difference. All problems were due to one parameter in datasource config - missing `persistent = false`. As I understood, when task is started, a `reminders` database connection was opened and then QueuedJobs table was trying to execute UPDATE using this existing connection and expectly failed because table is placed in another schema. Solution that I suggested is based on the fact that the schema is always added to the table name, so it does not matter in which connection the query is executed, so this workaround worked for me.
Anyway, I added `persistent = false` to my config and now task is completed and queued job status successfully updated.
